### PR TITLE
MINOR: Push down logic from TransactionManager to TxnPartitionEntry

### PR DESCRIPTION
**Pull request by ijuma**
And encapsulate TxnPartitionEntry state.

This makes it easier to understand the behavior and the paths through
which the state is updated.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/utils/ProducerIdAndEpoch.java [hunk 1]
```diff
@@ -35,7 +35,7 @@ public boolean isValid() {
 
     @Override
     public String toString() {
-        return "ProducerIdAndEpoch(producerId=" + producerId + ", epoch=" + epoch + ")";
+        return "(producerId=" + producerId + ", epoch=" + epoch + ")";
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/utils/ProducerIdAndEpoch.java:38]: `ProducerIdAndEpoch` doesn't add much value here, this is effectively a tuple.

**Comment by ijuma** [comment]
@jolshan I think this is ready for review.

**Comment by jolshan** [comment]
Sorry I lost track of this. I will take a look soon.

**Review by divijvaidya — COMMENTED** [review]
Dropped by to say thank you for this PR @ijuma. Small but critical step forward in the quest to restricting boundaries and interfaces in the code base.

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TxnPartitionEntry.java [hunk 2]
```diff
@@ -55,20 +63,111 @@ class TxnPartitionEntry {
...
+            int newSequence = inFlightBatch.baseSequence() - recordCount;
+            if (newSequence < 0)
+                throw new IllegalStateException("Sequence number for batch with sequence " + inFlightBatch.baseSequence()
+                        + " for partition " + topicPartition + " is going to become negative: " + newSequence);
+
+            inFlightBatch.resetProducerState(new ProducerIdAndEpoch(inFlightBatch.producerId(), inFlightBatch.producerEpoch()), newSequence);
+        });
+    }
+
+    int maybeUpdateLastAckedSequence(int sequence) {
```
Comment (jolshan) [inline clients/src/main/java/org/apache/kafka/clients/producer/internals/TxnPartitionEntry.java]: nit: for these two methods: in TransactionManager, they were listed before adjustSequencesDueToFailedBatch. Not a big deal, but was curious if we wanted the ordering to be consistent.  (Looks like we are pretty consistent with the TxnPartitionMap ordering)

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 3]
```diff
@@ -3449,7 +3449,8 @@ MAX_REQUEST_SIZE, ACKS_ALL, MAX_RETRIES, new SenderMetricsRegistry(new Metrics(t
         transactionManager.handleCompletedBatch(tp1b3, t1b3Response);
 
         assertFalse(transactionManager.hasInflightBatches(tp1));
-        assertEquals(1, transactionManager.sequenceNumber(tp1).intValue());
+        assertEquals(1, transactionManager.sequenceNumber(tp1));
+        assertEquals(1, transactionManager.sequenceNumber(tp1));
```
Comment (jolshan) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java]: nit: duplicated line?

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java [hunk 4]
```diff
@@ -778,22 +778,22 @@ public void testIdempotenceWithMultipleInflights() throws Exception {
         prepareAndReceiveInitProducerId(producerId, Errors.NONE);
         assertTrue(transactionManager.hasProducerId());
 
-        assertEquals(0, transactionManager.sequenceNumber(tp0).longValue());
+        assertEquals(0, transactionManager.sequenceNumber(tp0));
```
Comment (jolshan) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:781]: why were these longs in the first place?

Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/SenderTest.java:781]: I can't think of a good reason.

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java (hunk 3 above)
Comment (ijuma) [reply clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java]: Good catch!

File: clients/src/main/java/org/apache/kafka/clients/producer/internals/TxnPartitionEntry.java (hunk 2 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/producer/internals/TxnPartitionEntry.java]: I tried to have non private methods first. No particular reason for these two, I moved them above `adjustSequencesDueToFailedBatch`.

**Review by jolshan — APPROVED** [review]
Thanks! Will let build complete and check tests.

**Comment by ijuma** [comment]
JDK 8 build passed, the rest had a number of unrelated flaky failures:

> Build / JDK 21 and Scala 2.13 / kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest.testCreateTokenForOtherUserFails(String).quorum=kraft
Build / JDK 21 and Scala 2.13 / kafka.zk.ZkMigrationIntegrationTest.[2] Type=ZK, MetadataVersion=3.5-IV2, Security=PLAINTEXT
Build / JDK 21 and Scala 2.13 / org.apache.kafka.streams.integration.ConsistencyVectorIntegrationTest.shouldHaveSamePositionBoundActiveAndStandBy
Build / JDK 21 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()
Build / JDK 17 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBootstrapZkMigrationRecord()
Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.integration.NamedTopologyIntegrationTest.shouldAddNamedTopologyToRunningApplicationWithSingleInitialNamedTopology()
Build / JDK 17 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()
Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.ExactlyOnceSourceIntegrationTest.testSeparateOffsetsTopic
Build / JDK 11 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testNoOpRecordWriteAfterTimeout()
Build / JDK 11 and Scala 2.13 / org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest.testSendOffsetsWithGroupId(String).quorum=zk
Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.ConsistencyVectorIntegrationTest.shouldHaveSamePositionBoundActiveAndStandBy
Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.EOSUncleanShutdownIntegrationTest.shouldWorkWithUncleanShutdownWipeOutStateStore[exactly_once_v2]
Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.EosIntegrationTest.shouldWriteLatestOffsetsToCheckpointOnShutdown[at_least_once, processing threads = false]
Build / JDK 11 and Scala 2.13 / org.apache.kafka.streams.integration.EosIntegrationTest.shouldWriteLatestOffsetsToCheckpointOnShutdown[exactly_once, processing threads = false]
Build / JDK 11 and Scala 2.13 / org.apache.kafka.tools.MetadataQuorumCommandTest.[6] Type=Raft-Isolated, Name=testDescribeQuorumStatusSuccessful, MetadataVersion=3.7-IV1, Security=PLAINTEXT

**Pull request merged**